toolchain/semantics/semantics_parse_tree_handler.cpp

```diff
@@ -453,21 +474,63 @@ auto SemanticsParseTreeHandler::HandlePackageLibrary(
 
 auto SemanticsParseTreeHandler::HandleParameterList(ParseTree::Node parse_node)
     -> void {
-  // TODO: This should transform into a usable parameter list. For now
-  // it's unused and only stored so that node counts match.
-  node_stack_.PopAndDiscardSoloParseNode(ParseNodeKind::ParameterListStart);
-  node_stack_.Push(parse_node);
+  // If the final child node is a parameter, then we need to pop the node
+  // block for the final parameter off the stack.
+  if (auto parse_kind = parse_tree_->node_kind(node_stack_.PeekParseNode());
+      parse_kind != ParseNodeKind::ParameterListStart &&
+      parse_kind != ParseNodeKind::ParameterListComma) {
+    node_block_stack_.pop_back();
+  }
+
+  while (true) {
+    switch (auto parse_kind =
+                parse_tree_->node_kind(node_stack_.PeekParseNode())) {
+      case ParseNodeKind::ParameterListStart:
+        node_stack_.PopAndDiscardSoloParseNode(
+            ParseNodeKind::ParameterListStart);
+        node_stack_.Push(parse_node, node_block_vector_stack_.pop_back_val());
+        return;
+
+      case ParseNodeKind::ParameterListComma:
+        node_stack_.PopAndDiscardSoloParseNode(
+            ParseNodeKind::ParameterListComma);
+        break;
+
+      case ParseNodeKind::PatternBinding:
+        node_stack_.PopAndDiscardId(ParseNodeKind::PatternBinding);
+        break;
+
+      default:
+        // This should only occur for invalid parse trees.
+        CARBON_FATAL() << "TODO: " << parse_kind;
+    }
+  }
+
+  llvm_unreachable("loop always exits");
 }
 
 auto SemanticsParseTreeHandler::HandleParameterListComma(
-    ParseTree::Node /*parse_node*/) -> void {
-  CARBON_FATAL() << "TODO";
+    ParseTree::Node parse_node) -> void {
+  node_block_stack_.pop_back();
+  node_stack_.Push(parse_node);
+
+  // Possibly add a node block for the next parameter.
+  if (!NextParseNodeIs({ParseNodeKind::ParameterList})) {
```

Comment:
Using lookahead here feels unclean to me. I wonder if it would make sense to instead unconditionally create a new node block, and then discard it when we finish the parameter list if there was no additional parameter following the comma?

Reply:
Per discussion, this could come up with metaprogramming-related things. Removing NextParseNodeIs, I'll be working around it.